rusk/src/lib/chain/rusk.rs

```diff
@@ -495,15 +506,58 @@ fn execute(
     session: &mut Session,
     tx: &PhoenixTransaction,
 ) -> Result<CallReceipt<Result<Vec<u8>, ContractError>>, PiecrustError> {
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        let bytecode_charge =
+            deploy.bytecode.bytes.len() as u64 * GAS_PER_DEPLOY_BYTE;
+        if tx.payload().fee.gas_limit < bytecode_charge {
+            return Err(PiecrustError::Panic(
+                "not enough gas to deploy".into(),
+            ));
+        }
+    }
+
+    let tx_stripped = strip_off_bytecode(tx);
     // Spend the inputs and execute the call. If this errors the transaction is
     // unspendable.
     let mut receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
         TRANSFER_CONTRACT,
         "spend_and_execute",
-        tx,
+        tx_stripped.as_ref().unwrap_or(tx),
         tx.payload().fee.gas_limit,
```

Comment:
I already reduce the gas limit when doing the deployment proper (see line 548), yet for `spend_and_execute` I have a doubt, as here, gas limit distorts the `receipt.gas_limit`, which is then used for charging in case of OutOfGas (in line 565). Should we follow your suggestion here, we'd need to eliminate the use of `receipt.gas_limit` (at least for deploy txs, see below). On the contrary, if we kept using `receipt.gas_limit`, then we would distort the penalty, as it will be smaller than expected. `receipt.gas_limit` is used by the infrastructure if there is a contract call so we need to use it. On the other hand, in this case, we have no call, yet making such exception will complicate the code. So the penalty limit would need to be either the original limit if this is a deploy transaction or a `receipt.gas_limit` otherwise. I am not sure we want that.

Reply:
I think my question is partly due to my [other comment](https://github.com/dusk-network/rusk/pull/1941/files/0ecc2be4d84c1e36a89eb18859e4ea7253c21f60#r1681823448).

I was having a bit of trouble understanding what happens in the case of a contract deployment error. I was worried one or more assumptions that the code makes about `fn execute` were broken - mainly that an error on exit means an unspendable transaction.

I now understand that this isn't the case and that we neatly charge the transaction regardless of the outcome of deployment.

I would appreciate you updating the comment above `fn execute` to reflect the changes, and if possible extracting the deployment into a different function. It might make the code clearer.

Reply:
refactored deployment execution as separate method in rusk.rs